### PR TITLE
Validate certificates when verifying signed data.

### PR DESCRIPTION
**Pull request by SanjayVas**
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40)

https://rally1.rallydev.com/#/?detail=/task/672131373175&fdp=true

### CODE REVIEW

**Comment by SanjayVas** [comment]
Depends on https://github.com/world-federation-of-advertisers/common-jvm/pull/172

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
> Reviewed 7 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40)* status: 7 of 9 files reviewed, 3 unresolved discussions (waiting on @duliomatos1 and @SanjayVas)

___
*[`src/test/kotlin/org/wfanet/measurement/consent/client/dataprovider/DataProviderClientTest.kt` line 137 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40#-NI4Znxg4fqpevyNCfEk:-NI4ZnxhDzq1QAU3y_Is:b2btpvq) ([raw file](https://github.com/world-federation-of-advertisers/consent-signaling-client/blob/18a4ceffca545e84607fbaebdc022c608fdf3f02/src/test/kotlin/org/wfanet/measurement/consent/client/dataprovider/DataProviderClientTest.kt#L137)):*
> ```Kotlin
> 
>   @Test
>   fun `verifiesElgamalPublicKey does not throw exception when signature is valid`() = runBlocking {
> ```

needs a sad case that catches the two different error types
___
*[`src/test/kotlin/org/wfanet/measurement/consent/client/duchy/DuchyClientTest.kt` line 204 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40#-NI4ZtCu0eDao6yPsHty:-NI4ZtCu0eDao6yPsHtz:b-8b9sok) ([raw file](https://github.com/world-federation-of-advertisers/consent-signaling-client/blob/18a4ceffca545e84607fbaebdc022c608fdf3f02/src/test/kotlin/org/wfanet/measurement/consent/client/duchy/DuchyClientTest.kt#L204)):*
> ```Kotlin
> 
>   @Test
>   fun `verifiesElgamalPublicKey does not throw exception when signature is valid`() = runBlocking {
> ```

ditto on the tests
___
*[`src/test/kotlin/org/wfanet/measurement/consent/client/measurementconsumer/MeasurementConsumerClientTest.kt` line 177 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40#-NI4ZzCoA4d_MebcvcJD:-NI4ZzCoA4d_MebcvcJE:b-7ropln) ([raw file](https://github.com/world-federation-of-advertisers/consent-signaling-client/blob/18a4ceffca545e84607fbaebdc022c608fdf3f02/src/test/kotlin/org/wfanet/measurement/consent/client/measurementconsumer/MeasurementConsumerClientTest.kt#L177)):*
> ```Kotlin
> 
>   @Test
>   fun `verifyResult does not throw exception when signature is valid`() = runBlocking {
> ```

ditto

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
> Reviewed 18 of 18 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40)* status: all files reviewed, 3 unresolved discussions (waiting on @duliomatos1 and @SanjayVas)

___
*[`src/main/kotlin/org/wfanet/measurement/consent/client/common/VerificationException.kt` line 21 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40#-NI8YqVd3l1PYU0upGlO:-NI8YqVd3l1PYU0upGlP:bv8r4x0) ([raw file](https://github.com/world-federation-of-advertisers/consent-signaling-client/blob/0452c168abeaa78b7a8ed376900432b39b7fd722/src/main/kotlin/org/wfanet/measurement/consent/client/common/VerificationException.kt#L21)):*
> ```Kotlin
> import java.security.GeneralSecurityException
> 
> /** Consent signaling verification [Exception]. */
> ```

verification seems pretty broad. For example, SignatureException feels like a subclass of VerificationException. What's your intention by introducing VerificationException?

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40)* status: all files reviewed, 3 unresolved discussions (waiting on @duliomatos1 and @stevenwarejones)

___
*[`src/main/kotlin/org/wfanet/measurement/consent/client/common/VerificationException.kt` line 21 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40#-NI8YqVd3l1PYU0upGlO:-NI9g7Vu0xLixQQs7zjT:b-k68d2f) ([raw file](https://github.com/world-federation-of-advertisers/consent-signaling-client/blob/0452c168abeaa78b7a8ed376900432b39b7fd722/src/main/kotlin/org/wfanet/measurement/consent/client/common/VerificationException.kt#L21)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

verification seems pretty broad. For example, SignatureException feels like a subclass of VerificationException. What's your intention by introducing VerificationException?
</blockquote></details>

Due to the fact that some verification methods now do more than one thing (e.g. validating a certificate path in addition to verifying a signature), this PR changes those methods to throw throw exceptions rather than returning a boolean. This carries more information to the caller. Given that calling code treats these all as exceptional behavior, this is fine compared to e.g. returning a more detailed status object. This has the additional benefit that the verification methods can just pass through exceptions raised by `CertPathValidator` without modification.

`VerificationException` is added to cover cases where there isn't an exception type that already exists to cover the failure case (e.g. nonce mismatch). `CertPathValidatorException` and `SignatureException` are both existing exception types from the Java standard library. All three exceptions extend `GeneralSecurityException`.

In the PR that updates the calling code (https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/773), we sometimes wrap the other two exceptions in VerificationException as well to bubble them up after doing intermediate handling. `VerificationException` instances thrown by a verification method from this library are passed through in this case.

Basically: VerificationException is an exception for when you don't need more specific data in relation to a verification failure.

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40)* status: all files reviewed, 3 unresolved discussions (waiting on @duliomatos1 and @stevenwarejones)

___
*[`src/main/kotlin/org/wfanet/measurement/consent/client/common/VerificationException.kt` line 21 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40#-NI8YqVd3l1PYU0upGlO:-NI9wXUo7nI762t_ekKj:b-7k56l2) ([raw file](https://github.com/world-federation-of-advertisers/consent-signaling-client/blob/0452c168abeaa78b7a8ed376900432b39b7fd722/src/main/kotlin/org/wfanet/measurement/consent/client/common/VerificationException.kt#L21)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Due to the fact that some verification methods now do more than one thing (e.g. validating a certificate path in addition to verifying a signature), this PR changes those methods to throw throw exceptions rather than returning a boolean. This carries more information to the caller. Given that calling code treats these all as exceptional behavior, this is fine compared to e.g. returning a more detailed status object. This has the additional benefit that the verification methods can just pass through exceptions raised by `CertPathValidator` without modification.

`VerificationException` is added to cover cases where there isn't an exception type that already exists to cover the failure case (e.g. nonce mismatch). `CertPathValidatorException` and `SignatureException` are both existing exception types from the Java standard library. All three exceptions extend `GeneralSecurityException`.

In the PR that updates the calling code (https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/773), we sometimes wrap the other two exceptions in VerificationException as well to bubble them up after doing intermediate handling. `VerificationException` instances thrown by a verification method from this library are passed through in this case.

Basically: VerificationException is an exception for when you don't need more specific data in relation to a verification failure.
</blockquote></details>

I'm going ahead and adding more specific exceptions instead.

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40)* status: all files reviewed, 3 unresolved discussions (waiting on @duliomatos1 and @stevenwarejones)

___
*[`src/main/kotlin/org/wfanet/measurement/consent/client/common/VerificationException.kt` line 21 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40#-NI8YqVd3l1PYU0upGlO:-NI9xtnb2OkgORLKxdxr:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/consent-signaling-client/blob/0452c168abeaa78b7a8ed376900432b39b7fd722/src/main/kotlin/org/wfanet/measurement/consent/client/common/VerificationException.kt#L21)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

I'm going ahead and adding more specific exceptions instead.
</blockquote></details>

Done.

**Review by SanjayVas — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40)* status: 10 of 21 files reviewed, 3 unresolved discussions (waiting on @duliomatos1 and @stevenwarejones)

___
*[`src/test/kotlin/org/wfanet/measurement/consent/client/dataprovider/DataProviderClientTest.kt` line 137 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40#-NI4Znxg4fqpevyNCfEk:-NIA7wdg9Hhng7dTQ-cn:boj1qgt) ([raw file](https://github.com/world-federation-of-advertisers/consent-signaling-client/blob/18a4ceffca545e84607fbaebdc022c608fdf3f02/src/test/kotlin/org/wfanet/measurement/consent/client/dataprovider/DataProviderClientTest.kt#L137)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

needs a sad case that catches the two different error types
</blockquote></details>

Good catch. Done.
___
*[`src/test/kotlin/org/wfanet/measurement/consent/client/measurementconsumer/MeasurementConsumerClientTest.kt` line 177 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40#-NI4ZzCoA4d_MebcvcJD:-NIA7uIa38Rd2xgR2m-F:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/consent-signaling-client/blob/18a4ceffca545e84607fbaebdc022c608fdf3f02/src/test/kotlin/org/wfanet/measurement/consent/client/measurementconsumer/MeasurementConsumerClientTest.kt#L177)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

**Review by stevenwarejones — APPROVED** [review]
> Reviewed 11 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40)* status: all files reviewed (commit messages unreviewed), all discussions resolved (waiting on @duliomatos1)

**Review by duliomatos1 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40#-:-NIDQXWICFjl7MSIwUcM:bnfp4nl)* status: all files reviewed (commit messages unreviewed), all discussions resolved (waiting on @SanjayVas)

**Review by SanjayVas — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/consent-signaling-client/40)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @SanjayVas)
